File: Book Coding/Placement series AC/reverselinkedlist.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;




class Node{
    public:

    int data;
    Node* next;

    Node(int val){
        data = val;
        next = NULL;
    }
};




void insertAtTail(Node* &head, int val){
    Node* n = new Node(val);
    if (head == NULL){
        head = n;
        return;
    }

    Node *temp = head;
    while (temp->next != NULL){
        temp = temp->next;
    }
    temp->next = n;

}


void reverseList(Node* &head){
    Node* prev = NULL;
    Node* current = head;
    Node* nexty = head->next;

    while (current != NULL){
        if(current->next == NULL){
            current->next = prev;
            prev = current;
            current = nexty;
        }else{
            current->next = prev;
            prev = current;
            current = nexty;
            nexty = nexty->next;
        }
    }
    head = prev;
}



void printList(Node* head){
    if (head == NULL){
        cout << "The linked list is empty \n";
    }
    Node* temp = head;
    while (head!=NULL){
        cout << head->data << " -> ";
        head = head->next;
    }
    cout << "NULL \n";
    

}


int main(){
    Node* head = NULL;
    insertAtTail(head, 3);
    insertAtTail(head, 12);
    insertAtTail(head, 3);
    insertAtTail(head, 56);
    insertAtTail(head, 2);
    insertAtTail(head, 323);
    insertAtTail(head, 156);
    printList(head);
    reverseList(head);
    printList(head);



    return 0;
}

File: Book Coding/Placement series AC/reverse_linkedlist.py


class Node:
    def __init__(self, data):
        self.data = data 
        self.next = None 



class Linkedlist:
    def __init__(self):
        self.head = None 

    def pushattail(self, new_data):
        if self.head == None:
            self.head = Node(new_data)
        else:
            temp = self.head
            while temp.next is not None:
                temp = temp.next
            new_node = Node(new_data)
            temp.next = new_node


    def pushathead(self, new_data):
        temp = self.head
        new_node = Node(new_data)
        new_node.next = temp
        self.head = new_node

    def printlist(self):
        temp = self.head
        while temp:
            print(temp.data, end=" -> ")
            temp = temp.next
        print("")


    def reverse(self):
        prev = None 
        current = self.head
        nexty = self.head.next
        while current is not None:
            if current.next is not None:
                current.next = prev
                prev = current
                current = nexty
                nexty = nexty.next
            else:
                current.next = prev
                prev = current
                current = nexty

        self.head = prev


def reverse(head):
    if head == None or head.next == None:
        return head
    new_head = reverse(head.next)
    head.next.next = head
    # print(head.next.next.data)
    # print(head.next.data)
    head.next = None

    return new_head
    
    






if __name__ == '__main__':
    linky = Linkedlist()
    linky.pushattail(2)
    linky.pushattail(4)
    linky.pushattail(7)
    linky.pushattail(13)
    linky.printlist()
    linky.pushathead(21)
    linky.pushathead(45)
    linky.printlist()
    linky.head = reverse(linky.head)
    linky.printlist()
    linky.reverse()
    linky.printlist()





